external/fv3fit/fv3fit/keras/_models/shared/clip.py

```diff
@@ -0,0 +1,185 @@
+        concat_layers.append(pad_start)
+
+    concat_layers.append(layer)
+
+    if n_removed_from_end > 0:
+        pad_end = _zero_slice_with_placeholder_dim(layer, n_removed_from_end)
+        concat_layers.append(pad_end)
+
+    return tf.concat(concat_layers, axis=-1)
+
+
+def zero_fill_clipped_layers(
+    config: ClipConfig,
+    clipped_layers: Sequence[tf.Tensor],
+    variable_names: Sequence[str],
+    original_feature_dim_sizes: Sequence[int],
+) -> Sequence[tf.Tensor]:
+    # Takes layers and restores those that were clipped to their
+    # original feature dim sizes by replacing clipped levels with zeroes.
+    outputs = []
+    for layer, name, original_size in zip(
+        clipped_layers, variable_names, original_feature_dim_sizes
+    ):
+        if name in config.clip:
+            n_removed_from_start, n_removed_from_end = config.removed_sizes(
+                original_size, name
+            )
+            outputs.append(
+                zero_pad_output_feature_dim(
+                    layer, n_removed_from_start, n_removed_from_end
+                )
+            )
+        else:
+            outputs.append(layer)
+    return outputs
+
+
+def clip_layers(
+    config: ClipConfig, layers: Sequence[tf.Tensor], variable_names: Sequence[str]
+) -> Sequence[tf.Tensor]:
+    # Takes layers and applies clipping to those that have entries in the ClipConfig.
+    # The layer_names input is the list of variable names corresponding
+    # to each layer in layers.
+    # Returns a sequence of layers of the same length as input layers.
+    outputs = []
+    for layer, name in zip(layers, variable_names):
+        if name in config.clip:
+            outputs.append(config.clip_layer_along_feature_dim(layer, name))
+        else:
+            outputs.append(layer)
+    return outputs
+
+
+def clip_arrays(
+    config: ClipConfig, arrays: Sequence[np.ndarray], variable_names: Sequence[str],
+) -> Sequence[np.ndarray]:
+    # Takes np arrays and applies clipping to those that have entries
+    # in the ClipConfig. The variable_names input is the list of variable
+    # names corresponding to each element in arrays.
+    # Returns a sequence of arrays of the same length as input arrays
+    # Applies clipping to arrays alont the last (feature) dimension if
+    # they have an entry in clip config.
+    outputs = []
+    for array, name in zip(arrays, variable_names):
+        if name in config.clip:
+            slice = config.single_feature_slice(name)
+            start = slice.start or 0
+            stop = slice.stop or array.shape[-1]
+            outputs.append(
+                array.take(
+                    indices=range(start, stop), axis=-1  # type: ignore
+                )
+            )
+        else:
+            outputs.append(array)
+    return outputs
+
+
+def _zero_slice_with_placeholder_dim(layer: tf.Tensor, n_zero_levels: int) -> tf.Tensor:
```

Comment:
If there is a more concise way to create a zeroes layer with a placeholder first dimension I would love to hear it. Stackoverflow seems to suggest using `tf.placeholder` but that is not available in the version of Tensorflow that we use.